lib/ClangImporter/ClangImporter.cpp

```diff
@@ -7471,6 +7471,80 @@ bool importer::isForeignReferenceTypeWithoutImmortalAttrs(const clang::QualType
          !hasImmortalAtts(pointeeType->getDecl());
 }
 
+// Returns the given declaration along with all its parent declarations that are
+// reference types.
+llvm::SmallVector<const clang::RecordDecl *, 4>
+getRefParentDecls(const clang::RecordDecl *decl) {
+  assert(decl && "decl is null inside getRefParentDecls");
+
+  llvm::SmallVector<const clang::RecordDecl *, 4> matchingDecls;
+
+  if (hasImportAsRefAttr(decl)) {
+    matchingDecls.push_back(decl);
+  }
+
+  if (const auto *cxxRecordDecl = llvm::dyn_cast<clang::CXXRecordDecl>(decl)) {
+    if (!cxxRecordDecl->hasDefinition())
+      return matchingDecls;
+    cxxRecordDecl->forallBases([&](const clang::CXXRecordDecl *baseDecl) {
+      if (hasImportAsRefAttr(baseDecl)) {
+        matchingDecls.push_back(baseDecl);
+      }
+      return true;
+    });
+  }
+
+  return matchingDecls;
+}
+
+const clang::RecordDecl *getRefParentOrDiag(
+    const clang::RecordDecl *decl,
+    std::optional<std::reference_wrapper<ClangImporter::Implementation>>
```

Comment:
I wonder if a simple pointer that can be null is simpler than an optional with a reference wrapper.

Reply:
I wanted use modern C++ idioms but I can use pointer as well if you think that is safer or more performant.

Reply:
I think a pointer in this case is much easier to work with and better matches the code style in Swift.